Fix wrong next/previous lesson retrieved in presence of unpublished l…

eri-trabiccolo opened the pull request:
…essons

## Description
Fixes #1294 

Also fixed cs.

Running new tests I realized that `get_previous_lesson()` on the first lesson of a course returned NULL (actually no return), so I took this occasion to ensure it always returns false if no previous lesson is found, plus I slightly cleaned that method + his twin `get_nex_lesson()`.

## How has this been tested?
Old and new ad-hoc tests, plus some manual :P 

## Screenshots <!-- if applicable -->

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [x] My code has been tested.
- [x] My code passes all existing automated tests. <!-- Check code: `composer run-script tests-run`, Guidelines: https://github.com/gocodebox/lifterlms/blob/trunk/tests/README.md -->
- [x] My code follows the LifterLMS Coding & Documentation Standards. <!-- Check code: `composer run-script check-cs-errors`, Guidelines: https://github.com/gocodebox/lifterlms/blob/trunk/docs/coding-standards.md and https://github.com/gocodebox/lifterlms/blob/trunk/docs/documentation-standards.md -->

eri-trabiccolo committed:
Fix wrong next/previous lesson retrieved in presence of unpublished lessons

thomasplevy commented:
@eri-trabiccolo This *fixes the immediate problem identified by the bug report and looks good, however, in testing it I've come up with two new (pre-existing) but related issues that I think we can fix alongside this:

1) If there are unpublished lessons in a course, the course syllabus doesn't display them but it still counts them when displaying their order.

![Screenshot_2020-08-20_15-23-04](https://user-images.githubusercontent.com/1290739/90831681-1da04c00-e2f9-11ea-93f7-d139ce0e86fd.png)

This feels complicated to fix since the order is hard-coded as meta data. If we're to switch to generate the order dynamically that means that we need to run a lot of DB queries to be figuring out where it is. This is possibly a scenario that doesn't even really matter and maybe we can just call this "Expected" behavior and let users know that if they *do not want this to happen* they should be detaching draft lessons from a course.

2) In testing this I made the first lesson of section 2 a draft. Navigating from the last lesson of section 1 to the first available (actually second) lesson in section 2 worked but then navigating backwards (getting the previous lesson) didn't work.

The issue is:

If the first lesson of a section is drafted, I would expect the previous lesson of the second lesson in that section to be the last lesson of the previous section but it's actually returning false (as in there's no previous lesson).

eri-trabiccolo committed:
Fix previous lesson for the following scenario:
If the first lesson of a section is drafted, I would expect the previous
lesson of the second lesson in that section to be the last lesson of the
previous section but it's actually returning false (as in there's no previous lesson).

eri-trabiccolo committed:
Fix unpublished lessons in course's syllabus still counted when displaying lesssons order.

eri-trabiccolo commented:
@thomasplevy 
1. I think this should not be a big problem, and no I agree I don't want to run another query just for that.
Since we already pass the total_lessons to the lesson-preview template, I think we can pass the current section's lesson order as well.
Fixed in https://github.com/gocodebox/lifterlms/pull/1299/commits/ec5cf0f844f5af21e399916b6b3a798595fcb1c1
what do you think?
I kept the fall back on `$lesson->get_order()` if the progressive order is not passed, you know one might have overridden the syllabus template but not the lesson-preview.
2. I totally forgot to test that case which was exactly the mirror of the issue reported here. Basically I planned to do it but got distracted by the NULL issue in the `get_previous()` method and forgot about it. Sorry. Fixed in https://github.com/gocodebox/lifterlms/pull/1299/commits/11952c241ed0f632895a28242f7cb5de51dc3cd9

thomasplevy commented:
> @thomasplevy
> 
> 1. I think this should not be a big problem, and no I agree I don't want to run another query just for that.
>    Since we already pass the total_lessons to the lesson-preview template, I think we can pass the current section's lesson order as well.
>    Fixed in [ec5cf0f](https://github.com/gocodebox/lifterlms/commit/ec5cf0f844f5af21e399916b6b3a798595fcb1c1)
>    what do you think?
>    I kept the fall back on `$lesson->get_order()` if the progressive order is not passed, you know one might have overridden the syllabus template but not the lesson-preview.

Oh great. Every elegant!

> 2. I totally forgot to test that case which was exactly the mirror of the issue reported here. Basically I planned to do it but got distracted by the NULL issue in the `get_previous()` method and forgot about it. Sorry. Fixed in [11952c2](https://github.com/gocodebox/lifterlms/commit/11952c241ed0f632895a28242f7cb5de51dc3cd9)

Looks good!
